packages/app/cypress/e2e/runner/support/mochaEventsUtils.ts

```diff
@@ -0,0 +1,189 @@
+  }
+
+  return obj
+}
+
+const eventCleanseMap = {
+  snapshots: stringifyShort,
+  parent: stringifyShort,
+  tests: stringifyShort,
+  commands: stringifyShort,
+  invocationDetails: stringifyShort,
+  body: () => '[body]',
+  wallClockStartedAt: () => 'match.date',
+  lifecycle: () => 'match.number',
+  fnDuration: () => 'match.number',
+  duration: () => 'match.number',
+  afterFnDuration: () => 'match.number',
+  wallClockDuration: () => 'match.number',
+  stack: () => 'match.string',
+  file: (arg: string) => arg ? 'relative/path/to/spec.js' : undefined,
+  message: () => '[error message]',
+  sourceMappedStack: () => 'match.string',
+  parsedStack: () => 'match.array',
+  name: (n: string) => n === 'Error' ? 'AssertionError' : n,
+  err: () => {
+    return {
+      message: '[error message]',
+      name: 'AssertionError',
+      stack: 'match.string',
+      sourceMappedStack: 'match.string',
+      parsedStack: 'match.array',
+    }
+  },
+  startTime: new Date(0),
+  start: () => 'match.date',
+  end: () => 'match.date',
+  timings: (arg: MochaLifecycleData): SanitizedMochaLifecycleData => {
+    let sanitizedLifecycleData: SanitizedMochaLifecycleData = {
+      lifecycle: 'match.number',
+    }
+
+    for (const hook of hooks.before) {
+      const hooksToSanitize = arg[hook]
+
+      if (!hooksToSanitize) {
+        continue
+      }
+
+      sanitizedLifecycleData[hook] = hooksToSanitize.map((oldHook) => {
+        return {
+          hookId: oldHook.hookId,
+          fnDuration: 'match.number',
+          afterFnDuration: 'match.number',
+        }
+      })
+    }
+
+    if (arg.test) {
+      sanitizedLifecycleData.test = {
+        fnDuration: 'match.number',
+        afterFnDuration: 'match.number',
+      }
+    }
+
+    for (const hook of hooks.after) {
+      const hooksToSanitize = arg[hook]
+
+      if (!hooksToSanitize) {
+        continue
+      }
+
+      sanitizedLifecycleData[hook] = hooksToSanitize.map((oldHook) => {
+        return {
+          hookId: oldHook.hookId,
+          fnDuration: 'match.number',
+          afterFnDuration: 'match.number',
+        }
+      })
+    }
+
+    return sanitizedLifecycleData
+  },
+}
+
+const keysToEliminate = ['codeFrame', '_testConfig'] as const
+
+function removeUnusedKeysForTestSnapshot<T> (obj: T): T {
+  for (const key of keysToEliminate) {
+    delete obj[key]
+  }
+
+  for (const [key, value] of Object.entries(obj)) {
+    if (key in obj) {
+      const transform = eventCleanseMap[key]?.(value)
+
+      if (!transform) {
+        continue
+      }
+
+      obj[key] = transform
+    } else {
+      delete obj[key]
+    }
+  }
+
+  return obj
+}
+
+declare global {
+  interface Window {
+    bus: EventEmitter
+  }
+}
+
+class SnapshotError extends Error {
+  constructor (message: string) {
+    super()
+    this.message = `\n${message}`
+  }
+}
+
+export function runCypressInCypressMochaEventsTest<T> (snapshots: T, snapToCompare: keyof T, done: Mocha.Done) {
```

Comment:
This is the function that maps to the Excalidraw diagram in the PR description. Basically, the "inner" Cypress emits events, and the "outer" Cypress catches them via an `EventEmitter`. Once we encounter the final event, `mocha:end`, we run the assertion.